programs/util.c

```diff
@@ -633,6 +638,237 @@ const char* UTIL_getFileExtension(const char* infilename)
    return extension;
 }
 
+static int pathnameHas2Dots(const char *pathname)
+{
+    return NULL != strstr(pathname, "..");
+}
+
+int UTIL_isFileNameValidForMirroredOutput(const char *filename)
+{
+    return !pathnameHas2Dots(filename);
+}
+
+
+#define DIR_DEFAULT_MODE 0755
+static mode_t getDirMode(const char *dirName)
+{
+    struct stat st;
+    int ret = UTIL_getDirectoryStat(dirName, &st);
+    if (!ret) {
+        UTIL_DISPLAY("zstd: failed to get DIR stats %s: %s\n", dirName, strerror(errno));
+        return DIR_DEFAULT_MODE;
+    }
+    return st.st_mode;
+}
+
+static int makeDir(const char *dir, mode_t mode)
+{
+    int ret = mkdir(dir, mode);
+    if (ret != 0) {
+        if (errno == EEXIST)
+            return 0;
+        UTIL_DISPLAY("zstd: failed to create DIR %s: %s\n", dir, strerror(errno));
+    }
+    return ret;
+}
+
+void UTIL_convertPathnameToDirName(char *pathname)
+{
+    // get dir name from pathname similar to 'dirname'
+    assert(pathname != NULL);
+
+    // remove trailing '/' chars
+    int len = strlen(pathname);
+    while (pathname[len] == PATH_SEP) {
+        pathname[len] = '\0';
+        len--;
+    }
+    if (len == 0) return;
+
+    // "xyz/abc/file.txt" => "xyz/abc"
+    // "./file.txt" => "."
+    // "file.txt" => "."
+    char* pos = strrchr(pathname, PATH_SEP);
+    if (pos == NULL) {
+        pathname[0] = '.';
+        pathname[1] = '\0';
+    } else {
+        *pos = '\0';
+    }
+}
+
+static char* trimLeadingRootChar(char *filename)
+{
+    if ((filename != NULL) && (filename[0] == PATH_SEP))
+        return filename + 1;
+    return filename;
+}
+
+static char*
+trimLeadingCurrentDir(char *filename)
+{
+    if ((filename != NULL) && (filename[0] == '.') && (filename[1] == PATH_SEP))
```

Comment:
Is it valid to invoke this function with a `NULL` pointer ?
If not, you can make it a pre-condition of the function, document it, and just `assert` it in the function body.

Reply:
done.